operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java

```diff
@@ -252,6 +259,36 @@ private boolean isIntrospectionNeeded(Packet packet) {
               || isIntrospectVersionChanged(packet);
       }
 
+      private String introspectReason(Packet packet) {
+        StringBuilder sb = new StringBuilder("introspection needed because ");
+        if (getDomainTopology() == null) {
+          sb.append("domain topology is null");
+        } else if (isBringingUpNewDomain(packet)) {
+          sb.append("bringing up new domain");
+        } else {
+          sb.append("something else");
+        }
+        return sb.toString();
+      }
+
+      private String introspectionNotNeededReason(Packet packet) {
+        StringBuilder sb = new StringBuilder("introspection not needed because ");
+        if (getNumRunningServers() != 0) {
```

Comment:
The introspection may not be needed for different reasons even when there are running servers. For e.g. if the `domain.spec.introspectVersion` or  `domain.spec.image` value didn't change. Would the debug message be incorrect in such scenarios?

Reply:
Well, it's intended for developers to understand what is happening in the code, so I probably didn't handle every possible case, nor is that the goal. It should be easy enough to add cases as needed. If this were intended to end-users, I'd agree that we need to be more careful about not printing debug messages.

Reply:
ok. I was thinking that the debug message can just print the values for different conditions rather than checking for those conditions. For e.g.

```
StringBuilder sb = new StringBuilder("introspection not needed : ");
sb.append("Number of running servers: " + getNumRunningServers() + ",");
sb.append("Creating new servers: " + creatingServers(info) + ",");
sb.append("domain generation = ").append(getDomainGeneration())
                .append(" and packet last generation is ").append(packet.get(INTROSPECTION_DOMAIN_SPEC_GENERATION));
return sb.toString(); 
```
Anyways, I'm fine with refining and adding those cases later.